### ISSUE TITLE
Investigate alternative translations for inequality comparison

### ISSUE THREAD

**Issue by ranma42** [dotnet/efcore#34165]
The current translation of `nullableA != B` is `nullableA <> B OR nullableA IS NULL`.
This causes the duplication of the `nullableA` expression.
An alternative translation that could avoid this issue is `COALESCE(nullableA <> B, TRUE)`.

**Comment by ranma42** [comment]
Note that this would not fix the issue of double evaluation on SqlServer unless #32519 is fixed.

**Comment by ranma42** [comment]
⚠️ while this alternative translation is interesting to avoid duplication of sub-expressions, it might cause worse plans to be generated, for example when `nullableA` is a simple column expression.

**Comment by ranma42** [comment]
> For example Sqlite [could use indexes](https://www.sqlite.org/optoverview.html#evaluating_or_constraints_separately_and_taking_the_union_of_the_result) on a predicates like `nullableA <> B OR nullableA IS NULL` [...]

The index would not be useable for the `nullableA <> B`, so for the inequality it looks like a safe transformation.
For equality, more care might be needed. The same holds true for other comparisons; see https://github.com/dotnet/efcore/pull/33757/commits/4db69e0b3178c83db081d358302a0b7c8737edd2 where the same `CASE` structure is used (and the same issues with indexes would be hit).

**Comment by roji** [comment]
> The current translation of nullableA == B is nullableA = B AND nullableA IS NOT NULL.

This only referes to "non-optimized" mode, right?

> on the bright side, that translation is only used when !optimize, aka outside of predicates).

Note that !optimize does occur in predicate, e.g. within negation.

In any case, doesn't the switch to `IS NOT DISTINCT FROM` help here?

**Comment by ranma42** [comment]
> > The current translation of nullableA == B is nullableA = B AND nullableA IS NOT NULL.
> 
> This only referes to "non-optimized" mode, right?

Yes (for the `!=`/`OR` case, instead, we always emit the `nullableA <> B OR nullableA IS NULL`).

> 
> > on the bright side, that translation is only used when !optimize, aka outside of predicates).
> 
> Note that !optimize does occur in predicate, e.g. within negation.
> 
> In any case, doesn't the switch to `IS NOT DISTINCT FROM` help here?

Yes, `IS NOT DISTINCT FROM` is a better solution, but requires explicit support from the DB provider.
